src/extensions/qdatabase.c

```diff
@@ -86,13 +95,204 @@
 #define Q_MYSQL_OPT_CONNECT_TIMEOUT    (10)
 #define Q_MYSQL_OPT_READ_TIMEOUT       (30)
 #define Q_MYSQL_OPT_WRITE_TIMEOUT      (30)
-#endif
+#endif /* ENABLE_MYSQL */
+
+#ifdef ENABLE_PGSQL
+#include "libpq-fe.h"
+
+/*
+Wrap the low-level functions of the libpq library.
+Provide interfaces with naming convention similar to mysql_xyz().
+*/
+
+typedef enum {
+    PQ_UNKNOWN,
+    PQ_PING,
+    PQ_QUERY,
+    PQ_UPDATE,
+    PQ_TRANSACTION,
+} pgquery_t;
+
+typedef struct {
+    char   *emsg;
+    PGconn *pgconn;
+
+    PGresult *pgresult;
+    int rows;
+    int cols;
+    int cursor;
+} pgsql_t;
+
+static inline pgsql_t* pgsql_init(void)
+{
+    pgsql_t* pgsql = (pgsql_t *)calloc(1, sizeof(pgsql_t));
+    return pgsql;
+}
+
+static inline const char* pgsql_set_emsg(pgsql_t* pgsql, const char* emsg)
+{
+    if (pgsql == NULL || emsg == NULL) {
+        return NULL;
+    }
+
+    if (pgsql->emsg) {
+        free(pgsql->emsg);
+    }
+
+    pgsql->emsg = strdup(emsg);
+    return pgsql->emsg;
+}
+
+static inline void pgsql_close(pgsql_t* pgsql)
+{
+    if (pgsql == NULL) {
+        return;
+    }
+
+    if (pgsql->emsg) {
+        free(pgsql->emsg);
+        pgsql->emsg = NULL;
+    }
+
+    if (pgsql->pgresult) {
+        PQclear(pgsql->pgresult);
+        pgsql->pgresult = NULL;
+    }
+
+    if (pgsql->pgconn) {
+        PQfinish(pgsql->pgconn);
+        pgsql->pgconn = NULL;
+    }
+
+    free(pgsql);
+}
+
+static inline void pgsql_query_result(pgsql_t* pgsql)
+{
+    if (pgsql->pgresult) {
+        PQclear(pgsql->pgresult);
+        pgsql->pgresult = NULL;
+    }
+}
+
+static inline int pgsql_query(pgsql_t* pgsql, bool fetchtype,
+                                const char *query, pgquery_t type)
+{
+    if (pgsql == NULL || pgsql->pgconn == NULL) {
+        return -1; /* error */
+    }
+
+    /* free previous query results */
+    pgsql_query_result(pgsql);
+
+    if (type == PQ_PING) {
+        query = "";
+    }
+
+    PGresult *result = NULL;
+    ExecStatusType status = -1;
+
+    if (fetchtype && type == PQ_QUERY) {
+        if (PQsendQuery(pgsql->pgconn, query)) {
+            if (PQsetSingleRowMode(pgsql->pgconn)) {
+                result = PQgetResult(pgsql->pgconn);
+            } else {
+                return -1; /* error */
+            }
+        } else {
+            return -1; /* error */
+        }
+    } else {
+        result = PQexec(pgsql->pgconn, query);
+    }
+
+    if (result == NULL) {
+        return -1; /* error */
```

Comment:
We return -1 for all the error cases, how are users supposed to determine whether query was a problem, network failure or simply no matching rows, ... ?

Reply:
Detailed error information can be obtained by calling the get_error function